src/NuGet.Core/NuGet.Common/InProcLock.cs

```diff
@@ -0,0 +1,156 @@
+
+            await _dictionaryLock.WaitAsync(token);
+            try
+            {
+                lockState = GetOrCreate(key);
+            }
+            finally
+            {
+                _dictionaryLock.Release();
+            }
+
+            try
+            {
+                await lockState.Semaphore.WaitAsync(token);
+            }
+            catch
+            {
+                Interlocked.Decrement(ref lockState.Count);
+            }
+        }
+
+        internal void Enter(string key, CancellationToken token)
+        {
+            LockState lockState;
+
+            _dictionaryLock.Wait(token);
+            try
+            {
+                lockState = GetOrCreate(key);
+            }
+            finally
+            {
+                _dictionaryLock.Release();
+            }
+
+            try
+            {
+                lockState.Semaphore.Wait(token);
+            }
+            catch
+            {
+                Interlocked.Decrement(ref lockState.Count);
+            }
+        }
+
+        private LockState GetOrCreate(string key)
+        {
+            if (!_locks.TryGetValue(key, out var lockState))
+            {
+                lockState = new LockState();
+                lockState.Semaphore = new SemaphoreSlim(1);
+                lockState.Count = 1;
+                _locks[key] = lockState;
+            }
+            else
+            {
+                Interlocked.Increment(ref lockState.Count);
+            }
+
+            return lockState;
+        }
+
+        internal async Task ExitAsync(string key)
+        {
+            await _dictionaryLock.WaitAsync();
+            try
+            {
+                Cleanup(key);
+            }
+            finally
+            {
+                _dictionaryLock.Release();
+            }
+        }
+
+        internal void Exit(string key)
+        {
+            _dictionaryLock.Wait();
+            try
+            {
+                Cleanup(key);
+            }
+            finally
+            {
+                _dictionaryLock.Release();
+            }
+        }
+
+        private void Cleanup(string key)
+        {
+            var lockState = _locks[key];
+            lockState.Semaphore.Release();
+            var count = Interlocked.Decrement(ref lockState.Count);
+            if (count == 0)
+            {
+                lockState.Semaphore.Release();
+                _locks.Remove(key);
+            }
+        }
+
+        private class LockState
+        {
+            public SemaphoreSlim Semaphore;
+            public int Count;
+        }
+
+        #region IDisposable Support
+        private bool _disposedValue = false; // To detect redundant calls
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_disposedValue)
+            {
+                if (disposing)
+                {
+                    _dictionaryLock.Dispose();
+                    _locks.Clear();
```

Comment:
What if `_locks` contains outstanding locks?

Reply:
Good question. It's academic, since this class is `internal`, so shouldn't be used outside this assembly, and the only instance of it is static, so will never get disposed. But is it acceptable to block `Dispose` while it waits for child locks to become free? Should it throw an `InvalidOperationException`, saying can't dispose while at least one lock is still in use?

If this was a reusable class, would be the expected behaviour be?